rootfs/scheduler/__init__.py

```diff
@@ -1420,5 +1506,292 @@ def get_node(self, name, **kwargs):
+                pods = self.get_pods(namespace, labels=labels).json()
+                for pod in pods['items']:
+                    # Get more information on why a pod is pending
+                    if pod['status']['phase'] == 'Pending':
+                        reason, message = self._pod_pending_status(pod)
+                        # If pulling an image is taking long then increase the timeout
+                        timeout += self._handle_pod_long_image_pulling(pod, reason)
+
+                        # handle errors and bubble up if need be
+                        self._handle_pod_image_errors(pod, reason, message)
+
+                logger.info("waited {}s and {} pods are in service".format(waited, availablePods))
+
+            waited += 1
+            time.sleep(1)
+
+    def _build_deployment_manifest(self, namespace, name, image, command, **kwargs):
+        replicas = kwargs.get('replicas', 0)
+        batches = kwargs.get('deploy_batches', None)
+        tags = kwargs.get('tags', {})
+
+        labels = {
+            'app': namespace,
+            'type': kwargs.get('app_type'),
+            'heritage': 'deis',
+        }
+
+        manifest = {
+            'kind': 'Deployment',
+            'apiVersion': 'extensions/v1beta1',
+            'metadata': {
+                'name': name,
+                'labels': labels,
+                'annotations': {
+                    'kubernetes.io/change-cause': kwargs.get('release_summary', '')
+                }
+            },
+            'spec': {
+                'replicas': replicas,
+                'selector': {
+                    'matchLabels': labels
+                }
+            }
+        }
+
+        # Add in Rollback (if asked for)
+        rollback = kwargs.get('rollback', False)
+        if rollback:
+            # http://kubernetes.io/docs/user-guide/deployments/#rollback-to
+            if rollback is True:
+                # rollback to the latest known working revision
+                revision = 0
+            elif isinstance(rollback, int) or isinstance(rollback, str):
+                # rollback to a particular revision
+                revision = rollback
+
+            # This gets cleared from the template after a rollback is done
+            manifest['spec']['rollbackTo'] = {'revision': str(revision)}
+
+        # Add deployment strategy
+
+        # see if application or global deploy batches are defined
+        maxSurge = self._get_deploy_steps(batches, tags)
+        # if replicas are higher than maxSurge then the old deployment is never scaled down
+        # maxSurge can't be 0 when maxUnavailable is 0 and the other way around
+        if replicas > 0 and replicas < maxSurge:
+            maxSurge = replicas
+
+        # http://kubernetes.io/docs/user-guide/deployments/#strategy
+        manifest['spec']['strategy'] = {
+            'rollingUpdate': {
+                'maxSurge': maxSurge,
+                # TODO potentially support this
```

Comment:
Coming up, had it written down on on a piece of paper. Transcribing to tickets now